flax/linen/linear.py

```diff
@@ -51,6 +54,38 @@ def _canonicalize_tuple(x: Union[Sequence[int], int]) -> Tuple[int, ...]:
     return (x,)
 
 
+def _canonicalize_dtypes(
+    input_dtype: InexactDType,
+    param_dtype: Optional[InexactDType],
+    computation_dtype: Optional[InexactDType]) -> Tuple[InexactDType,
+                                                        InexactDType]:
+  returned_param_dtype = input_dtype if param_dtype is None else param_dtype
+  dtype = (jnp.result_type(input_dtype, returned_param_dtype)
+           if computation_dtype is None else computation_dtype)
+
+  assert jnp.issubdtype(input_dtype, jnp.inexact)
+  if jnp.issubdtype(input_dtype, jnp.complexfloating):
+    assert jnp.issubdtype(returned_param_dtype, jnp.complexfloating)
+    assert jnp.issubdtype(dtype, jnp.complexfloating)
+  return returned_param_dtype, dtype
+
+
+def _canonicalize_numeric_dtypes(
```

Comment:
I think it would be nice to expose and document these APIs publicly so people can implement their own layers according to the spec. 

I think you can have a canonicalize and canonicalize_dtype_inexact.

Also I think you should use the true dtypes of the params and not the param_dtype because users can cast params after construction

Reply:
> I think it would be nice to expose and document these APIs publicly so people can implement their own layers according to the spec.

Yes, great point.

Did you want them in a new file?  I'll expose them here for now, but let me know if you want them somewhere else.

> Also I think you should use the true dtypes of the params 

Sorry, I'm not sure what you want me to do here.

Reply:
I would either make a dtypes.py or put it in module.py

> Sorry, I'm not sure what you want me to do here? 

You can pass the parameter values into jnp.result_type(*inputs, *params)

Reply:
the canonicalize_dtype should only infer the dtype if it is None. param_dtype defaults to float32 and cannot be inferred (so it's not Optional[Dtype] but Dtype). This way you can init the params and then use the params + inputs to infer the dtype

Reply:
> the canonicalize_dtype should only infer the dtype if it is None

Right, that's what it does.

> This way you can init the params and then use the params + inputs to infer the dtype

What is the difference between
- passing the parameter d-type to the canonicalize function, versus
- initializing the parameters using the parameter dtype, and passing the parameters to the canonicalize function?

The exact same type inference will happen either way.

Reply:
The param_dtype argument is not the same as the parameter dtype. In 99% of the cases users will stick to the default float32. For example during eval the two can diverge because a user does something like:
```
eval_params = jax.tree_map(lambda x: x.asarray(jnp.bfloat16), params)
```

The reasoning behind this common trick is that small updates during training don't accumulate well in half precision but during eval the weights are static and using half precision is a free gain in most cases.


The core idea behind canonicalize dtype is to reproduce what you normally expect from an equivalent numpy function. So imagine that we wrote nn.Dense as a pure numpy function:

```
def dense(input, kernel, bias, dtype=None):
  dtype = dtype or jnp.result_type(input, kernel, bias)
  return input @ kernel + bias
```

Here the user has to provide kernel and bias so the dtypes of the params are determined by the user and don't depend on input.dtype. The default dtype is just np.result_type(input, kernel, bias). 

The reason why we allow this dtype to be inferred is because it preserves the precision of the computation.
When inferring params however you are inferring behavior. e.g.: A complex linear mapping is something different than a linear real mapping even if you have a complex input. Similarly, a learned half precision dense does something very different in practice than a f32 dense layer even if the inputs are in half precision.

A second argument against a None default for param_dtype is that it would cause a big backwards incompatible change because many users really on f32 defaults for half precision inputs.

Reply:
> The reasoning behind this common trick is that small updates during training don't accumulate well in half precision but during eval the weights are static and using half precision is a free gain in most cases.

Okay!  I see what you're saying now.  I've started this work [here](https://github.com/NeilGirdhar/flax/blob/modified_dd/flax/linen/linear.py#L110), but it gets tricky [here](https://github.com/NeilGirdhar/flax/blob/modified_dd/flax/linen/attention.py#L281).  Hopefully, I'm on the right track now.

> A complex linear mapping is something different than a linear real mapping even if you have a complex input.

Yes that's true, but the parameter cotangents will still be complex, so I imagine that in most cases, complex inputs imply complex parameters.

> A second argument against a None default for param_dtype is that it would cause a big backwards incompatible change because many users really on f32 defaults for half precision inputs.

Right, that's why I switched all the defaults to float32.  This way, this PR just provides the ability to specify computation and parameter dtype.  It doesn't change behavior.

However, I still think you should deprecate this default in another PR.  As you say: "The core idea behind canonicalize dtype is to reproduce what you normally expect from an equivalent numpy function."  And the equivalent numpy function always produces outputs based on its input types.  It doesn't silently widen everything to float32.

Also, if someone does want half-precision (or complex, or double precision) computation throughtout their network, they will need to select it for every module.  I read the FLIP, but I think things can change in the future.  Dobule precision is already starting to be [as performant](https://blogs.nvidia.com/blog/2020/05/14/double-precision-tensor-cores/) as single precision on some GPUs.  It would be better not to bake in defaults that you can't easily remove.